src/ncp/rcp_host.cpp

```diff
@@ -438,12 +451,82 @@ const char *RcpHost::GetThreadVersion(void)
     return version;
 }
 
+static bool needRejoin(const otOperationalDatasetTlvs &aLhs, const otOperationalDatasetTlvs &aRhs)
+{
+    bool result = false;
+
+    otOperationalDataset lhsDataset;
+    otOperationalDataset rhsDataset;
+
+    SuccessOrExit(otDatasetParseTlvs(&aLhs, &lhsDataset));
+    SuccessOrExit(otDatasetParseTlvs(&aRhs, &rhsDataset));
+
+    result =
+        (lhsDataset.mChannel == rhsDataset.mChannel) &&
+        (memcmp(lhsDataset.mNetworkKey.m8, rhsDataset.mNetworkKey.m8, sizeof(lhsDataset.mNetworkKey)) == 0) &&
+        (memcmp(lhsDataset.mExtendedPanId.m8, rhsDataset.mExtendedPanId.m8, sizeof(lhsDataset.mExtendedPanId)) == 0);
+
+exit:
+    return result;
+}
+
 void RcpHost::Join(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs, const AsyncResultReceiver &aReceiver)
 {
-    OT_UNUSED_VARIABLE(aActiveOpDatasetTlvs);
+    otError                  error = OT_ERROR_NONE;
+    std::string              errorMsg;
+    bool                     receiveResultHere = true;
+    otOperationalDatasetTlvs curDatasetTlvs;
+
+    VerifyOrExit(mInstance != nullptr, error = OT_ERROR_INVALID_STATE, errorMsg = "OT is not initialized");
+    VerifyOrExit(mThreadEnabledState != ThreadEnabledState::kStateDisabling, error = OT_ERROR_BUSY,
+                 errorMsg = "Thread is disabling");
+    VerifyOrExit(mThreadEnabledState == ThreadEnabledState::kStateEnabled, error = OT_ERROR_INVALID_STATE,
+                 errorMsg = "Thread is not enabled");
 
-    // TODO: Implement Join under RCP mode.
-    mTaskRunner.Post([aReceiver](void) { aReceiver(OT_ERROR_NOT_IMPLEMENTED, "Not implemented!"); });
+    otbrLogInfo("Start joining...");
+
+    error = otDatasetGetActiveTlvs(mInstance, &curDatasetTlvs);
+    if (error == OT_ERROR_NONE && needRejoin(aActiveOpDatasetTlvs, curDatasetTlvs) && IsAttached())
+    {
+        // Do not leave and re-join if this device has already joined the same network. This can help elimilate
+        // unnecessary connectivity and topology disruption and save the time for re-joining. It's more useful for use
+        // cases where Thread networks are dynamically brought up and torn down (e.g. Thread on mobile phones).
+        SuccessOrExit(error    = otDatasetSetActiveTlvs(mInstance, &aActiveOpDatasetTlvs),
+                      errorMsg = "Failed to set Active Operational Dataset");
```

Comment:
If already joined to same network, do we need to set the active dataset again? 
This can trigger registration/resync of dataset with leader? Is this desired/interntional?

Reply:
Yes, this is out of consideration for unknown TLVs in the dataset. It's very difficult to check if two datasets are completely the same with the existence of unknown TLVs (because of the order). So we simply set the active dataset again. But we want to avoid rejoin as possible. Thoughts?

Reply:
I think it sounds okay to do so. The leader should combine them. This may add some extra traffic/exchanges when the parameters fully match, but I expect it should be harmless otherwise.